File: flutter_app/lib/signaling.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:path_provider/path_provider.dart';
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'package:url_launcher/url_launcher.dart';

typedef void StreamStateCallback(MediaStream stream);

class Signaling {
  Map<String, dynamic> configuration = {
    'iceServers': []
  };

  IO.Socket? socket;
  String? roomId;
  RTCPeerConnection? peerConnection;
  MediaStream? localStream;
  List<RTCIceCandidate> _candidateQueue = [];
  bool _remoteDescriptionSet = false;
  MediaStream? remoteStream;
  Function(MediaStream stream)? onAddRemoteStream;

  VoidCallback? onCallEnded;
  MediaRecorder? _mediaRecorder;

  // Replace with your machine's IP address if running on a real device
  // For Android Emulator use 10.0.2.2
  // For iOS Simulator use localhost
  String get serverUrl {
    if (kIsWeb) {
      return 'http://localhost:3000';
    } else if (defaultTargetPlatform == TargetPlatform.android) {
      return 'http://10.0.2.2:3000';
    }
    return 'http://localhost:3000';
  }

  void connect() {
    socket = IO.io(serverUrl, <String, dynamic>{
      'transports': ['websocket'],
      'autoConnect': false,
    });

    socket!.connect();

    socket!.onConnect((_) {
      print('Connected to signaling server');
    });

    socket!.on('ice_servers', (data) {
      print('Received ICE servers: $data');
      configuration['iceServers'] = data;
    });

    socket!.on('room_created', (data) async {
      print('Room $data created');
      roomId = data;
    });

    socket!.on('room_joined', (data) async {
      print('Room $data joined');
      roomId = data;
      _createOffer();
    });

    socket!.on('full_room', (data) {
      print('Room $data is full');
    });

    socket!.on('user_left', (data) {
      print('User left the room');
      onCallEnded?.call();
    });

    socket!.on('webrtc_offer', (data) async {
      print('Received offer');
      var sdp = RTCSessionDescription(data['sdp'], data['type']);
      await _createAnswer(sdp);
    });

    socket!.on('webrtc_answer', (data) async {
      print('Received answer');
      var sdp = RTCSessionDescription(data['sdp'], data['type']);
      await peerConnection?.setRemoteDescription(sdp);
      _remoteDescriptionSet = true;
      _processCandidateQueue();
    });

    socket!.on('webrtc_ice_candidate', (data) async {
      print('Received ICE candidate');
      // Fix for type error: ensure sdpMLineIndex is int
      int? sdpMLineIndex = data['sdpMLineIndex'] is int 
          ? data['sdpMLineIndex'] 
          : int.tryParse(data['sdpMLineIndex'].toString());
          
      var candidate = RTCIceCandidate(
        data['candidate'],
        data['sdpMid'],
        sdpMLineIndex,
      );
      
      if (_remoteDescriptionSet) {
        await peerConnection?.addCandidate(candidate);
      } else {
        _candidateQueue.add(candidate);
      }
    });
  }

  Future<void> _processCandidateQueue() async {
    for (var candidate in _candidateQueue) {
      await peerConnection?.addCandidate(candidate);
    }
    _candidateQueue.clear();
  }

  Future<void> openUserMedia(
      RTCVideoRenderer localVideo, RTCVideoRenderer remoteVideo) async {
    var stream = await navigator.mediaDevices
        .getUserMedia({'video': true, 'audio': true});

    localVideo.srcObject = stream;
    localStream = stream;

    remoteVideo.srcObject = await createLocalMediaStream('key');
  }

  Future<void> joinRoom(String roomId, RTCVideoRenderer remoteVideo) async {
    this.roomId = roomId;
    await _initializePeerConnection(remoteVideo);
    socket!.emit('join', roomId);
  }

  Future<void> _createOffer() async {
    RTCSessionDescription description =
        await peerConnection!.createOffer({'offerToReceiveVideo': 1});
    var session = parse(description.sdp!);
    print(json.encode(session));
    
    await peerConnection!.setLocalDescription(description);

    socket!.emit('webrtc_offer', {
      'type': 'offer',
      'sdp': description.sdp,
      'roomId': roomId,
    });
  }

  Future<void> _createAnswer(RTCSessionDescription remoteOffer) async {
    await peerConnection!.setRemoteDescription(remoteOffer);
    _remoteDescriptionSet = true;
    _processCandidateQueue();

    RTCSessionDescription description =
        await peerConnection!.createAnswer({'offerToReceiveVideo': 1});

    await peerConnection!.setLocalDescription(description);

    socket!.emit('webrtc_answer', {
      'type': 'answer',
      'sdp': description.sdp,
      'roomId': roomId,
    });
  }

  void registerPeerConnectionListeners() {
    peerConnection?.onIceGatheringState = (RTCIceGatheringState state) {
      print('ICE gathering state changed: $state');
    };

    peerConnection?.onConnectionState = (RTCPeerConnectionState state) {
      print('Connection state changed: $state');
    };

    peerConnection?.onSignalingState = (RTCSignalingState state) {
      print('Signaling state changed: $state');
    };

    peerConnection?.onIceCandidate = (RTCIceCandidate candidate) {
      print('ICE candidate generated');
      socket!.emit('webrtc_ice_candidate', {
        'candidate': candidate.candidate,
        'sdpMid': candidate.sdpMid,
        'sdpMLineIndex': candidate.sdpMLineIndex,
        'roomId': roomId,
      });
    };
  }
  
  // Helper to parse SDP (optional, for debugging)
  dynamic parse(String sdp) {
    return sdp; // Placeholder
  }

  void hangUp(RTCVideoRenderer localVideo, RTCVideoRenderer remoteVideo) {
    stopRecording();
    if (localStream != null) {
      localStream!.dispose();
    }
    if (remoteStream != null) {
      remoteStream!.dispose();
    }
    if (peerConnection != null) {
      peerConnection!.close();
    }
    if (socket != null) {
      if (roomId != null) {
        socket!.emit('leave_room', roomId);
      }
      // Do not disconnect the socket, so we can join again later
      // socket!.disconnect();
    }
    localVideo.srcObject = null;
    remoteVideo.srcObject = null;
    _candidateQueue.clear();
    _remoteDescriptionSet = false;
    roomId = null;
  }
  void onRemoteHangUp(RTCVideoRenderer remoteVideo) {
    stopRecording();
    if (remoteStream != null) {
      remoteStream!.dispose();
    }
    if (peerConnection != null) {
      peerConnection!.close();
    }
    // Do NOT disconnect socket or clear roomId
    // Just clear remote video and connection state
    remoteVideo.srcObject = null;
    _candidateQueue.clear();
    _remoteDescriptionSet = false;
    
    // Re-initialize peer connection for next call
    if (roomId != null) {
      _initializePeerConnection(remoteVideo);
    }
  }

  Future<void> startRecording(MediaStream stream) async {
    if (_mediaRecorder != null) return;
    
    print("Starting recording...");
    
    _mediaRecorder = MediaRecorder();
    
    if (kIsWeb) {
      // Web uses startWeb()
      _mediaRecorder!.startWeb(stream);
      print("Recording started on web");
    } else {
      // Mobile/Desktop uses start()
      final appDocDir = await getApplicationDocumentsDirectory();
      final videoDir = Directory('${appDocDir.path}/videos');
      if (!await videoDir.exists()) {
        await videoDir.create(recursive: true);
      }
      String filePath = '${videoDir.path}/video_call_${DateTime.now().millisecondsSinceEpoch}.mp4';
      
      await _mediaRecorder!.start(
        filePath, 
        videoTrack: stream.getVideoTracks().isNotEmpty ? stream.getVideoTracks().first : null,
      );
      print("Recording started at $filePath");
    }
  }

  Future<void> stopRecording() async {
    if (_mediaRecorder != null) {
      print("Stopping recording...");
      final result = await _mediaRecorder!.stop();
      print("Recording stopped: $result");
      
      if (kIsWeb && result != null) {
        // On web, result might be a blob URL
        // We can try to open it to trigger download/view
        if (result is String) {
          await launchUrl(Uri.parse(result));
        }
      }
      
      _mediaRecorder = null;
    }
  }

  Future<void> _initializePeerConnection(RTCVideoRenderer remoteVideo) async {
    peerConnection = await createPeerConnection(configuration);
    registerPeerConnectionListeners();
    localStream?.getTracks().forEach((track) {
      peerConnection?.addTrack(track, localStream!);
    });
    
    peerConnection?.onTrack = (RTCTrackEvent event) {
      print('Got remote track: ${event.streams[0]}');
      if (event.streams.isNotEmpty) {
        remoteVideo.srcObject = event.streams[0];
        onAddRemoteStream?.call(event.streams[0]);
        // Start recording when we receive the remote stream (2 people in room)
        startRecording(event.streams[0]);
      }
    };
  }
}

File: react_native_app/Signaling.js
import io from 'socket.io-client';
import {
    RTCPeerConnection,
    RTCIceCandidate,
    RTCSessionDescription,
    mediaDevices,
} from 'react-native-webrtc';
import Geolocation from 'react-native-geolocation-service';
import { Platform } from 'react-native';

export default class Signaling {
    constructor() {
        this.socket = null;
        this.roomId = null;
        this.peerConnection = null;
        this.localStream = null;
        this.remoteStream = null;
        this.candidateQueue = [];
        this.remoteDescriptionSet = false;

        // Callbacks
        this.onAddRemoteStream = null;
        this.onLocationReceived = null;
        this.onCallEnded = null;

        // Default config, updated by server
        this.configuration = {
            iceServers: [],
        };

        // Server URL
        // Android Emulator: 10.0.2.2, iOS Simulator: localhost
        this.serverUrl = Platform.OS === 'android'
            ? 'http://10.0.2.2:3000'
            : 'http://localhost:3000';
    }

    connect() {
        this.socket = io(this.serverUrl, {
            transports: ['websocket'],
            autoConnect: false,
        });

        this.socket.connect();

        this.socket.on('connect', () => {
            console.log('Connected to signaling server');
        });

        this.socket.on('ice_servers', (data) => {
            console.log('Received ICE servers:', data);
            this.configuration.iceServers = data;
        });

        this.socket.on('room_created', (data) => {
            console.log(`Room ${data} created`);
            this.roomId = data;
        });

        this.socket.on('room_joined', (data) => {
            console.log(`Room ${data} joined`);
            this.roomId = data;
            this._createOffer();
        });

        this.socket.on('full_room', (data) => {
            console.log(`Room ${data} is full`);
        });

        this.socket.on('user_left', () => {
            console.log('User left the room');
            if (this.onCallEnded) this.onCallEnded();
        });

        this.socket.on('webrtc_offer', async (data) => {
            console.log('Received offer');
            const sdp = new RTCSessionDescription(data);
            await this._createAnswer(sdp);
        });

        this.socket.on('webrtc_answer', async (data) => {
            console.log('Received answer');
            if (data.location) {
                console.log('Received location:', data.location);
                if (this.onLocationReceived) this.onLocationReceived(data.location);
            }
            const sdp = new RTCSessionDescription(data);
            await this.peerConnection.setRemoteDescription(sdp);
            this.remoteDescriptionSet = true;
            this._processCandidateQueue();
        });

        this.socket.on('webrtc_ice_candidate', async (data) => {
            console.log('Received ICE candidate');
            const candidate = new RTCIceCandidate(data);
            if (this.remoteDescriptionSet) {
                await this.peerConnection.addIceCandidate(candidate);
            } else {
                this.candidateQueue.push(candidate);
            }
        });
    }

    async _processCandidateQueue() {
        for (const candidate of this.candidateQueue) {
            await this.peerConnection.addIceCandidate(candidate);
        }
        this.candidateQueue = [];
    }

    async openUserMedia(localRenderSetter, remoteRenderSetter) {
        try {
            const stream = await mediaDevices.getUserMedia({
                audio: true,
                video: true,
            });

            // Callback to set local stream URL/object for display
            // Note: react-native-webrtc RTCView uses streamURL for older versions or object for newer
            // Assuming newer version based on install
            this.localStream = stream;
            localRenderSetter(stream);

            // Remote stream placeholder is usually handled when track is received
        } catch (err) {
            console.error('Error opening user media:', err);
        }
    }

    async joinRoom(roomId, remoteRenderSetter) {
        this.roomId = roomId;
        // Pass the setter so we can update UI when track arrives
        await this._initializePeerConnection(remoteRenderSetter);
        this.socket.emit('join', roomId);
    }

    async _initializePeerConnection(remoteRenderSetter) {
        this.peerConnection = new RTCPeerConnection(this.configuration);

        // Add local tracks
        if (this.localStream) {
            this.localStream.getTracks().forEach(track => {
                this.peerConnection.addTrack(track, this.localStream);
            });
        }

        this.peerConnection.onicecandidate = (event) => {
            if (event.candidate) {
                console.log('ICE candidate generated');
                this.socket.emit('webrtc_ice_candidate', {
                    candidate: event.candidate.candidate,
                    sdpMid: event.candidate.sdpMid,
                    sdpMLineIndex: event.candidate.sdpMLineIndex,
                    roomId: this.roomId,
                });
            }
        };

        this.peerConnection.ontrack = (event) => {
            console.log('Got remote track:', event.streams[0]);
            if (event.streams && event.streams.length > 0) {
                this.remoteStream = event.streams[0];
                remoteRenderSetter(event.streams[0]);
                if (this.onAddRemoteStream) this.onAddRemoteStream(event.streams[0]);

                // Start recording when remote stream is added (assuming 2 participants)
                this.startRecording();
            }
        };
    }

    async _createOffer() {
        const offer = await this.peerConnection.createOffer({ offerToReceiveVideo: 1 });
        await this.peerConnection.setLocalDescription(offer);

        this.socket.emit('webrtc_offer', {
            type: 'offer',
            sdp: offer.sdp,
            roomId: this.roomId,
        });
    }

    async _createAnswer(remoteOffer) {
        await this.peerConnection.setRemoteDescription(remoteOffer);
        this.remoteDescriptionSet = true;
        this._processCandidateQueue();

        const answer = await this.peerConnection.createAnswer({ offerToReceiveVideo: 1 });
        await this.peerConnection.setLocalDescription(answer);

        let locationString = null;
        try {
            const position = await this._determinePosition();
            locationString = `Lat: ${position.coords.latitude}, Lng: ${position.coords.longitude}`;
        } catch (e) {
            console.log('Error getting location:', e);
        }

        this.socket.emit('webrtc_answer', {
            type: 'answer',
            sdp: answer.sdp,
            roomId: this.roomId,
            location: locationString,
        });
    }

    _determinePosition() {
        return new Promise((resolve, reject) => {
            Geolocation.getCurrentPosition(
                (position) => {
                    resolve(position);
                },
                (error) => {
                    reject(error);
                },
                { enableHighAccuracy: true, timeout: 15000, maximumAge: 10000 }
            );
        });
    }

    hangUp(localRenderSetter, remoteRenderSetter) {
        this.stopRecording();

        if (this.peerConnection) {
            this.peerConnection.close();
            this.peerConnection = null;
        }

        if (this.socket) {
            if (this.roomId) {
                this.socket.emit('leave_room', this.roomId);
            }
            // Do not disconnect socket
        }

        this.candidateQueue = [];
        this.remoteDescriptionSet = false;
        this.roomId = null;

        // Clear streams ui
        // localStream remains active for local view typically, unless we want to stop it
        // Flutter implementation closes tracks. 
        if (this.localStream) {
            this.localStream.getTracks().forEach(t => t.stop());
            this.localStream = null;
        }

        localRenderSetter(null);
        remoteRenderSetter(null);
    }

    onRemoteHangUp(remoteRenderSetter) {
        this.stopRecording();

        if (this.peerConnection) {
            this.peerConnection.close();
        }

        // Do not disconnect socket
        this.candidateQueue = [];
        this.remoteDescriptionSet = false;
        remoteRenderSetter(null);

        // Re-init happens when joining again or handled by caller logic
        // In Flutter we re-init PC immediately. Here we can wait for join call or do it lightly.
        // For simplicity, we just clean up. `joinRoom` calls `_initializePeerConnection` anyway.
    }

    startRecording() {
        if (Platform.OS === 'web' && window.MediaRecorder && this.remoteStream) {
            try {
                this.mediaRecorder = new MediaRecorder(this.remoteStream);
                this.recordedChunks = [];

                this.mediaRecorder.ondataavailable = (event) => {
                    if (event.data && event.data.size > 0) {
                        this.recordedChunks.push(event.data);
                    }
                };

                this.mediaRecorder.start();
                console.log('Recording started');
            } catch (e) {
                console.error('Error starting recording:', e);
            }
        }
    }

    stopRecording() {
        if (this.mediaRecorder && this.mediaRecorder.state !== 'inactive') {
            this.mediaRecorder.stop();
            console.log('Recording stopped');

            // On web, define what to do on stop (e.g., save/download)
            if (Platform.OS === 'web') {
                this.mediaRecorder.onstop = () => {
                    const blob = new Blob(this.recordedChunks, { type: 'video/webm' });
                    const url = URL.createObjectURL(blob);
                    const a = document.createElement('a');
                    a.style.display = 'none';
                    a.href = url;
                    a.download = `recording_${new Date().toISOString()}.webm`;
                    document.body.appendChild(a);
                    a.click();
                    URL.revokeObjectURL(url);
                    console.log('Recording saved');
                };
            }
            this.mediaRecorder = null;
        }
    }
}
